### ISSUE TITLE
PKI - Add missing parameters in multiple resources

### ISSUE THREAD

**Issue by sestegra** [hashicorp/terraform-provider-vault#1510]
In general, the following scenarios are not supported:
- the managed keys by a KMS
- the new issuer API (from Vault 1.11)

Regarding [official documentation](https://developer.hashicorp.com/vault/api-docs/secret/pki#generate-root), following parameters or possible values are missing:

For generate root and intermediate
- `type`: the possible values `existing` and `kms` are missing
- `issuer_name`
- `key_name`
- `key_ref`
- `managed_key_name`
- `managed_key_id`
- `not_before_duration`
- `not_after``

For sign and role
- `issuer_ref`

**Comment by benashz** [comment]
Hi @sestegra ,

The managed keys feature will be coming out in the 3.9.0 release. As for the new Issuer PKI support, that should come out soon after 3.9.0.

Thanks,

Ben

**Comment by krarey** [comment]
In addition to adding support for `issuer_ref` on the noted resources, we should also add a resource for calling the `root/replace` and `config/issuers` endpoints to update the associated root CA certificate and default issuer associations: https://developer.hashicorp.com/vault/api-docs/secret/pki?optInFrom=vault-io#set-issuers-configuration

**Comment by krarey** [comment]
Similarly, `vault_pki_secret_backend_config_ca` should return the `imported_issuers` attribute from the Vault response, to chain into the aforementioned issuer configuration.

**Comment by lboynton** [comment]
Hi @benashz, any plans to release the new Issuer PKI support soon? It would be great to be able to specify the default issuer through terraform.

**Comment by geekofalltrades** [comment]
The new issuer feature in Vault 1.11 is 7 months old. This lack of first-class support in the Terraform provider is embarrassing.

Here's what I've done to keep the pre-1.11 behavior in the PKI backend. You need at least Vault 1.11.6 or 1.12.2 for the [`default_follows_latest_issuer`](https://developer.hashicorp.com/vault/api-docs/secret/pki#default_follows_latest_issuer) option. If you don't have those versions, the option is silently ignored and you will have a bad time.

```hcl
resource "vault_mount" "pki" {
  path = "pki"
  type = "pki"
}

# XXX; The vault_pki_secret_backend_config_ca resource has no drift detection,
# as the API it wraps is write-only. If this resource persists in the state
# while Vault state is wiped underneath it, Terraform will not correctly
# detect that it needs to be recreated.
#
# lifecycle.replace_triggered_by on the config_ca doesn't solve this by itself,
# as when Vault state is wiped, vault_mount.pki is planned to be created, not
# updated or replaced, and resource creation doesn't trigger
# replace_triggered_by.
#
# To get the behavior we want, use this null_resource as a shim. It survives
# Vault state being wiped, and plans to replace itself if the accessor ID of the
# PKI mount changes. This replacement can be used to trigger
# replace_triggered_by on the config_ca resource.
resource "null_resource" "pki_backend_tracker" {
  triggers = {
    accessor = vault_mount.pki.accessor
  }
}

resource "vault_pki_secret_backend_config_ca" "pki" {
  lifecycle {
    replace_triggered_by = [null_resource.pki_backend_tracker]
  }

  backend    = vault_mount.pki.path
  pem_bundle = local.pki
}

# Vault 1.11 added new functionality in the pki backend in which there can be
# multiple issuing CAs ("issuers") installed in one mount. Adding a new CA cert
# now creates a new issuer, but doesn't update the backend to issue from that
# new issuer by default. The "default_follows_latest_issuer" option preserves
# the pre-1.11 behavior of always issuing from the last installed issuer.
#
# https://developer.hashicorp.com/vault/api-docs/secret/pki#notice-about-new-multi-issuer-functionality
# https://developer.hashicorp.com/vault/api-docs/secret/pki#default_follows_latest_issuer
#
# The config/issuers endpoint requires that we provide an issuer to set as the
# default in its "issuer" field, so first, figure out what the current default
# is and just pass it back. We can't do this until an issuer is actually
# installed, so we depends_on the pki config CA being installed. This is fine as
# long as we're not changing the CA in the same run we're first applying this
# config, as that would create a new issuer which isn't the default and then
# that issuer would get orphaned. Make sure this resource is applied on the
# current CA before rotating to a new CA.
data "vault_generic_secret" "pki_default_issuer" {
  depends_on = [vault_pki_secret_backend_config_ca.pki]

  path = "${vault_mount.pki.path}/issuer/default"
}

resource "vault_generic_endpoint" "pki_config_issuers" {
  path = "${vault_mount.pki.path}/config/issuers"
  data_json = jsonencode({
    default                       = data.vault_generic_secret.pki_default_issuer.data.issuer_id,
    default_follows_latest_issuer = true,
  })
  disable_delete = true
}
```

**Comment by ingvarch** [comment]
Any updates here ? We really need this feature. 
Still waiting.

**Comment by japtain-cack** [comment]
I am in need of these changes as well.

**Comment by camarrone** [comment]
The Vault provider should be in line with the vault API.

As said above by @itspngu, this is a huge pain point and my case it is a show stopper and a breaking point.

We can't keep writing alternative solutions and be more error prone due to lack of support on this provider.

Any ETC on this?

Thanks.

**Comment by vinay-gopalan** [comment]
Hey folks, apologies for the delayed response! 🙏🏼 

We 100% agree that the TFVP should be in line with Vault's API! While Vault's feature set development continues to scale with multiple teams, the TFVP team is also trying to improve upon the long term sustainability of this project and scale this project's development by collaborating with more teams and contributors. This collaborative effort has growing pains and is slow, so again we apologize and appreciate the patience. 🙇🏼 

In that context, the TFVP team is collaborating w/ the Vault CryptoSec team to appropriately support the multi-issuer feature set within the TFVP 😄  At present the development for this feature is on track w/ Vault's upcoming 1.14 release in mid-June; we'll try our best for an earlier release! I'll be sure to continue updating this thread w/ any relevant info, and thanks once again for the patience!

**Comment by nickschoenbaechler** [comment]
Is mid-June still the target? I'm trying to decide if it's best for me to work around this or to expect that TFVP will be ready in the next week or so.

**Comment by vinay-gopalan** [comment]
@nickschoenbaechler, yes, we are still targeting mid-June for the release of this feature. We're currently testing a feature branch on our end.

**Comment by vinay-gopalan** [comment]
Hey folks, just wanted to drop an update here that we released Multi-Issuer Functionality for PKI in the TFVP as part of [v3.17.0](https://github.com/hashicorp/terraform-provider-vault/releases/tag/v3.17.0). We apologize for the delay on the feature parity and appreciate your patience!

Going to close this issue for now, but please let us know of if you face any issues! Thanks 🙏🏼
